convert-hf-to-gguf.py

```diff
@@ -1781,6 +1791,25 @@ def write_tensors(self):
 class Qwen2Model(Model):
     model_arch = gguf.MODEL_ARCH.QWEN2
 
+    def __init__(self, dir_model: Path, ftype: int, fname_out: Path, is_big_endian: bool, use_temp_file: bool):
+        super().__init__(dir_model, ftype, fname_out, is_big_endian, use_temp_file)
+        self.hparams = Qwen2Model.load_hparams(dir_model)
+
+    @staticmethod
+    def load_hparams(dir_model):
+        with open(dir_model / "config.json", "r", encoding="utf-8") as f1, \
+            open(dir_model / "tokenizer_config.json", "r", encoding="utf-8") as f2:
+            hparams = json.load(f1)
+            hparams.update(json.load(f2))
+
+        return hparams
+    
+    def set_vocab(self):
+        if self.hparams.get("tokenizer_class") == "PreTrainedTokenizerFast":
+            self._set_vocab_sentencepiece()
+        else:
+            self._set_vocab_gpt2()
```

Comment:
_set_vocab_sentencepiece raises FileNotFoundError for non-llama vocabs, so you should rely on that instead of checking tokenizer_class.

Reply:
I think I have overriden it in Qwen2 so it won't affect others? Which model did u meet problems with?

It only happens for CodeQwen1.5, cuz it uses a different tokenizer compared with Qwen1.5

Reply:
> I haven't tried your code, I'm just suggesting that it should be consistent with other uses of _set_vocab_sentencepiece:
> 
> https://github.com/ggerganov/llama.cpp/blob/0d56246f4b9764158525d894b96606f6163c53a8/convert-hf-to-gguf.py#L1301-L1304
> 
> We happen to do it this way because of the deepseek models, which set tokenizer_class to LlamaTokenizerFast despite using a GPT2-style BPE tokenizer. _set_vocab_sentencepiece uses LlamaHfVocab which has some logic to identify the _actual_ tokenizer.

it is a bit different from this case actually. just the same model architecture but with different tokenizers. but i think your solution should also be ok, cuz for qwen1.5 we don't have the file tokenizer.model. i'll give a shot

Reply:
made it simplified following ur suggestions.